lib/icinga/dependency.cpp

```diff
@@ -9,106 +9,170 @@
 #include "base/exception.hpp"
 #include <map>
 #include <sstream>
+#include <unordered_map>
 #include <utility>
+#include <variant>
 
 using namespace icinga;
 
 REGISTER_TYPE(Dependency);
 
-bool Dependency::m_AssertNoCyclesForIndividualDeps = false;
+INITIALIZE_ONCE(Dependency::StaticInitialize);
 
-struct DependencyCycleNode
+void Dependency::StaticInitialize()
 {
-	bool Visited = false;
-	bool OnStack = false;
-};
-
-struct DependencyStackFrame
-{
-	ConfigObject::Ptr Node;
-	bool Implicit;
-
-	inline DependencyStackFrame(ConfigObject::Ptr node, bool implicit = false) : Node(std::move(node)), Implicit(implicit)
-	{ }
-};
-
-struct DependencyCycleGraph
-{
-	std::map<Checkable::Ptr, DependencyCycleNode> Nodes;
-	std::vector<DependencyStackFrame> Stack;
-};
-
-static void AssertNoDependencyCycle(const Checkable::Ptr& checkable, DependencyCycleGraph& graph, bool implicit = false);
+	dynamic_cast<ConfigType*>(Dependency::TypeInstance.get())->BeforeOnAllConfigLoaded.connect(BeforeOnAllConfigLoadedHandler);
+}
 
-static void AssertNoParentDependencyCycle(const Checkable::Ptr& parent, DependencyCycleGraph& graph, bool implicit)
+/**
+ * Helper class to search for cycles in the dependency graph.
+ *
+ * State is stored inside the class and no synchronization is done,
+ * hence instances of this class must not be used concurrently.
+ */
+class DependencyCycleChecker
 {
-	if (graph.Nodes[parent].OnStack) {
-		std::ostringstream oss;
-		oss << "Dependency cycle:\n";
-
-		for (auto& frame : graph.Stack) {
-			oss << frame.Node->GetReflectionType()->GetName() << " '" << frame.Node->GetName() << "'";
+	struct Node
+	{
+		bool Visited = false;
+		bool OnStack = false;
+		std::vector<Dependency::Ptr> ExtraDependencies;
+	};
+
+	// std::hash is only implemented starting from Boost 1.74, use a (ordered) map as fallback.
+	// https://github.com/boostorg/smart_ptr/commit/5a18ffdc5609a0e64b63e47cb81c4f0847e0c087
+#if BOOST_VERSION >= 107400
+	std::unordered_map<Checkable::Ptr, Node> m_Nodes;
+#else
+	std::map<Checkable::Ptr, Node> m_Nodes;
+#endif
```

Comment:
Why all these workarounds when you could just use raw pointers instead 😄?

Reply:
> just implementing std::hash<intrusive_ptr> for Boost < 1.74 within Icinga 2, maybe that would actually be the better idea

I agree.

Reply:
> just implementing std::hash<intrusive_ptr> for Boost < 1.74 within Icinga 2, maybe that would actually be the better idea

I agree.